app/controllers/application_controller/wait_for_task.rb

```diff
@@ -11,65 +11,64 @@ def wait_for_task
+      send(async_params[:action]) # call the orig. method
     end
   end
 
-  def browser_refresh_task(task_id, should_flash = false)
-    session[:async][:interval] += 250 if session[:async][:interval] < 5000 # Slowly move up to 5 second retries
+  def browser_refresh_task(task_id, async_params, async_interval, should_flash: false)
+    async_interval += 250 if async_interval < 5000 # Slowly move up to 5 second retries
     render :update do |page|
       page << javascript_prologue
-      ajax_call = remote_function(:url => {:action => 'wait_for_task', :task_id => task_id})
-      page << "setTimeout(\"#{ajax_call}\", #{session[:async][:interval]});"
+      ajax_call = remote_function(:url => {:action => 'wait_for_task', :task_id => task_id, :async_params => async_params, :async_interval => :async_interval})
+      page << "setTimeout(\"#{ajax_call}\", #{async_interval});"
       page.replace("flash_msg_div", :partial => "layouts/flash_msg") if should_flash
       page << "miqScrollTop();" if @flash_array.present?
     end
   end
   private :browser_refresh_task
 
   #
-  # :task_id => id of task to wait for
-  # :action  => 'action_to_call' -- action to be called when the task finishes
-  # :rx_action => 'method_to_call' -- a method to create a RxJs message
-  # :flash => true|false -- output queued flash messages *while waiting*
+  # @option options :task_id [Numeric] id of task to wait for
+  # @option options :action  [String] action to be called when the task finishes
+  # @option options :rx_action [String]  a method to create a RxJs message
+  # @option options :flash [Boolean] output queued flash messages *while waiting*
+  # @option options :extra_params [Hash] asynchronous 
   #
   def initiate_wait_for_task(options = {})
     task_id = options[:task_id]
-    session[:async] ||= {}
-    session[:async][:interval] ||= 1000 # Default interval to 1 second
-    session[:async][:params] ||= {}
+    async_interval = 1000 # Default interval to 1 second
 
     # save the incoming parms + extra_params
-    session[:async][:params] = params.to_unsafe_h.merge(options[:extra_params] || {})
-    session[:async][:params][:task_id] = task_id
+    async_params = params.to_unsafe_h.merge(options[:extra_params] || {})
+    async_params[:task_id] = task_id
 
     # override method to be called, when the task is done
-    session[:async][:params][:action] = options[:action] if options.key?(:action)
+    async_params[:action] = options[:action] if options.key?(:action)
 
     if options.key?(:rx_action)
       raise "Unsupported combination" if options.key?(:action)
 
-      session[:async][:params][:action] = 'wait_for_task_rx'
-      session[:async][:params][:rx_action] = options[:rx_action]
+      async_params[:action] = 'wait_for_task_rx'
+      async_params[:rx_action] = options[:rx_action]
     end
 
-    browser_refresh_task(task_id, !!options[:flash])
+    session[:async][:params] = async_params
```

Comment:
I think you need the `||= {}` up above to ensure this path exists.

Reply:
good eye.
My question below is if we can drop `session[:async]` all together.
In the meantime, I'll add it back.